File: tests/output/javascript/src/client/search.test.ts
/* eslint-disable @typescript-eslint/no-unused-vars, require-await */
// @ts-nocheck Failing tests will have type errors, but we cannot suppress them even with @ts-expect-error because it doesn't work for a block of lines.
import type { SearchClient } from '@experimental-api-clients-automation/client-search';
import { searchClient } from '@experimental-api-clients-automation/client-search';
import { echoRequester } from '@experimental-api-clients-automation/requester-node-http';
import type { EchoResponse } from '@experimental-api-clients-automation/requester-node-http';

const appId = 'test-app-id';
const apiKey = 'test-api-key';

function createClient(): SearchClient {
  return searchClient(appId, apiKey, { requester: echoRequester() });
}

describe('api', () => {
  test('calls api with correct read host', async () => {
    const $client = searchClient('test-app-id', 'test-api-key', {
      requester: echoRequester(),
    });

    const result = (await $client.get({
      path: '/test',
    })) as unknown as EchoResponse;

    expect(result.host).toEqual('test-app-id-dsn.algolia.net');
  });

  test('calls api with correct write host', async () => {
    const $client = searchClient('test-app-id', 'test-api-key', {
      requester: echoRequester(),
    });

    const result = (await $client.post({
      path: '/test',
    })) as unknown as EchoResponse;

    expect(result.host).toEqual('test-app-id.algolia.net');
  });
});

describe('commonApi', () => {
  test('calls api with correct user agent', async () => {
    const $client = createClient();

    const result = (await $client.post({
      path: '/test',
    })) as unknown as EchoResponse;

    expect(decodeURI(result.algoliaAgent)).toMatch(
      /^Algolia for JavaScript \(\d+\.\d+\.\d+(-.*)?\)(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-.*)?\))?)*(; Search (\(\d+\.\d+\.\d+(-.*)?\)))(; [a-zA-Z. ]+ (\(\d+((\.\d+)?\.\d+)?(-.*)?\))?)*$/
    );
  });

  test('calls api with default read timeouts', async () => {
    const $client = createClient();

    const result = (await $client.get({
      path: '/test',
    })) as unknown as EchoResponse;

    expect(result).toEqual(
      expect.objectContaining({ connectTimeout: 2000, responseTimeout: 5000 })
    );
  });

  test('calls api with default write timeouts', async () => {
    const $client = createClient();

    const result = (await $client.post({
      path: '/test',
    })) as unknown as EchoResponse;

    expect(result).toEqual(
      expect.objectContaining({ connectTimeout: 2000, responseTimeout: 30000 })
    );
  });
});

describe('parameters', () => {
  test('client throws with invalid parameters', async () => {
    try {
      const $client = searchClient('', '', { requester: echoRequester() });

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch('`appId` is missing.');
    }
    try {
      const $client = searchClient('', 'my-api-key', {
        requester: echoRequester(),
      });

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch('`appId` is missing.');
    }
    try {
      const $client = searchClient('my-app-id', '', {
        requester: echoRequester(),
      });

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch('`apiKey` is missing.');
    }
  });

  test('`addApiKey` throws with invalid parameters', async () => {
    const $client = createClient();

    try {
      const result = (await $client.addApiKey(null)) as unknown as EchoResponse;

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch(
        'Parameter `apiKey` is required when calling `addApiKey`.'
      );
    }
  });

  test('`addOrUpdateObject` throws with invalid parameters', async () => {
    const $client = createClient();

    try {
      const result = (await $client.addOrUpdateObject({
        objectID: 'my-object-id',
        body: {},
      })) as unknown as EchoResponse;

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch(
        'Parameter `indexName` is required when calling `addOrUpdateObject`.'
      );
    }
    try {
      const result = (await $client.addOrUpdateObject({
        indexName: 'my-index-name',
        body: {},
      })) as unknown as EchoResponse;

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch(
        'Parameter `objectID` is required when calling `addOrUpdateObject`.'
      );
    }
    try {
      const result = (await $client.addOrUpdateObject({
        indexName: 'my-index-name',
        objectID: 'my-object-id',
      })) as unknown as EchoResponse;

      throw new Error('test is expected to throw error');
    } catch (e) {
      expect((e as Error).message).toMatch(
        'Parameter `body` is required when calling `addOrUpdateObject`.'
      );
    }
  });
});

File: tests/output/java/src/test/java/com/algolia/client/search.test.java
package com.algolia.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.algolia.EchoInterceptor;
import com.algolia.EchoResponse;
import com.algolia.api.SearchClient;
import com.algolia.model.search.*;
import com.algolia.utils.ClientOptions;
import com.algolia.utils.HttpRequester;
import java.util.*;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class SearchClientClientTests {

  private HttpRequester requester;
  private EchoInterceptor echo;

  @BeforeAll
  void init() {
    requester = new HttpRequester();
    echo = new EchoInterceptor();
    requester.addInterceptor(echo.getEchoInterceptor());
  }

  SearchClient createClient() {
    return new SearchClient("appId", "apiKey", new ClientOptions().setRequester(requester));
  }

  @Test
  @DisplayName("calls api with correct read host")
  void apiTest0() {
    SearchClient $client = new SearchClient("test-app-id", "test-api-key", new ClientOptions().setRequester(requester));

    String path0 = "/test";

    $client.get(path0);
    EchoResponse result = echo.getLastResponse();

    assertEquals("test-app-id-dsn.algolia.net", result.host);
  }

  @Test
  @DisplayName("calls api with correct write host")
  void apiTest1() {
    SearchClient $client = new SearchClient("test-app-id", "test-api-key", new ClientOptions().setRequester(requester));

    String path0 = "/test";

    $client.post(path0);
    EchoResponse result = echo.getLastResponse();

    assertEquals("test-app-id.algolia.net", result.host);
  }

  @Test
  @DisplayName("calls api with correct user agent")
  void commonApiTest0() {
    SearchClient $client = createClient();

    String path0 = "/test";

    $client.post(path0);
    EchoResponse result = echo.getLastResponse();

    {
      String regexp =
        "^Algolia for Java \\(\\d+\\.\\d+\\.\\d+(-.*)?\\)(; [a-zA-Z. ]+" +
        " (\\(\\d+((\\.\\d+)?\\.\\d+)?(-.*)?\\))?)*(; Search" +
        " (\\(\\d+\\.\\d+\\.\\d+(-.*)?\\)))(; [a-zA-Z. ]+" +
        " (\\(\\d+((\\.\\d+)?\\.\\d+)?(-.*)?\\))?)*$";
      assertTrue(
        result.headers.get("user-agent").matches(regexp),
        "Expected " + result.headers.get("user-agent") + " to match the following regex: " + regexp
      );
    }
  }

  @Test
  @DisplayName("calls api with default read timeouts")
  void commonApiTest1() {
    SearchClient $client = createClient();

    String path0 = "/test";

    $client.get(path0);
    EchoResponse result = echo.getLastResponse();

    assertEquals(2000, result.connectTimeout);
    assertEquals(5000, result.responseTimeout);
  }

  @Test
  @DisplayName("calls api with default write timeouts")
  void commonApiTest2() {
    SearchClient $client = createClient();

    String path0 = "/test";

    $client.post(path0);
    EchoResponse result = echo.getLastResponse();

    assertEquals(2000, result.connectTimeout);
    assertEquals(30000, result.responseTimeout);
  }

  @Test
  @DisplayName("client throws with invalid parameters")
  void parametersTest0() {
    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          SearchClient $client = new SearchClient("", "", new ClientOptions().setRequester(requester));
        }
      );
      assertEquals("`appId` is missing.", exception.getMessage());
    }
    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          SearchClient $client = new SearchClient("", "my-api-key", new ClientOptions().setRequester(requester));
        }
      );
      assertEquals("`appId` is missing.", exception.getMessage());
    }
    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          SearchClient $client = new SearchClient("my-app-id", "", new ClientOptions().setRequester(requester));
        }
      );
      assertEquals("`apiKey` is missing.", exception.getMessage());
    }
  }

  @Test
  @DisplayName("`addApiKey` throws with invalid parameters")
  void parametersTest1() {
    SearchClient $client = createClient();

    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          ApiKey apiKey0 = null;

          $client.addApiKey(apiKey0);
          EchoResponse result = echo.getLastResponse();
        }
      );
      assertEquals("Parameter `apiKey` is required when calling `addApiKey`.", exception.getMessage());
    }
  }

  @Test
  @DisplayName("`addOrUpdateObject` throws with invalid parameters")
  void parametersTest2() {
    SearchClient $client = createClient();

    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          String indexName0 = null;
          String objectID0 = "my-object-id";
          Map<String, Object> body0 = new HashMap<>();
          {}

          $client.addOrUpdateObject(indexName0, objectID0, body0);
          EchoResponse result = echo.getLastResponse();
        }
      );
      assertEquals("Parameter `indexName` is required when calling `addOrUpdateObject`.", exception.getMessage());
    }
    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          String indexName0 = "my-index-name";
          String objectID0 = null;
          Map<String, Object> body0 = new HashMap<>();
          {}

          $client.addOrUpdateObject(indexName0, objectID0, body0);
          EchoResponse result = echo.getLastResponse();
        }
      );
      assertEquals("Parameter `objectID` is required when calling `addOrUpdateObject`.", exception.getMessage());
    }
    {
      Exception exception = assertThrows(
        Exception.class,
        () -> {
          String indexName0 = "my-index-name";
          String objectID0 = "my-object-id";
          Object body0 = null;

          $client.addOrUpdateObject(indexName0, objectID0, body0);
          EchoResponse result = echo.getLastResponse();
        }
      );
      assertEquals("Parameter `body` is required when calling `addOrUpdateObject`.", exception.getMessage());
    }
  }
}
